### PR TITLE
KBC-1868 add dataProvider to test default and dev branch

### PR DESCRIPTION
**Pull request by martinjunger**
Jira: KBC-1868

Before asking for review make sure that:

- [x] New client method(s) has tests
- [x] Apiary file is updated
- [x] You declared if there is a BC break or not (will affect next release of a client)

---

### CODE REVIEW

File: src/Keboola/StorageApi/Components.php [hunk 1]
```diff
@@ -25,6 +25,14 @@ public function __construct(Client $client)
         $this->client = $client;
     }
 
+    /**
+     * @return Client
+     */
+    public function getClient()
```
Comment (martinjunger) [inline src/Keboola/StorageApi/Components.php:31]: Abych se dostal ke klientovi z Components, musel jsem udelat **public** metodu. Vadi to necemu?

File: tests/Common/ComponentsTest.php [hunk 2]
```diff
@@ -1178,14 +1323,18 @@ public function testUpdateConfigWithoutIdShouldNotBeAllowed()
         }
     }
 
-
-    public function testComponentConfigsVersionsRollback()
+    // TODO Configuration main-1 not found
```
Comment (martinjunger) [inline tests/Common/ComponentsTest.php:1326]: Pomoci `// TODO chyba` jsou oznaceny ty, ktery se chovaji jinak.

File: tests/Common/ComponentsTest.php [hunk 3]
```diff
@@ -219,11 +256,18 @@ public function testComponentConfigRenew()
...
     }
 
-    public function testComponentConfigDelete()
+    /**
+     * @dataProvider componentsProvider
+     */
+    public function testComponentConfigDelete(callable $getComponents)
     {
+        if ($this->dataName() === 'devBranch') {
+            $this->markTestSkipped('Deleting configuration from trash is not allowed in development branches.');
```
Comment (martinjunger) [inline tests/Common/ComponentsTest.php:265]: Nektery testy pro devBranch maji `markTestSkipped`, protoze to nedava smysl.

File: tests/Common/ComponentsTest.php [hunk 4]
```diff
@@ -38,10 +38,39 @@ public function setUp()
...
+        ];
+    }
+
+    /**
+     * @dataProvider componentsProvider
+     */
+    public function testGetComponentDetail(callable $getComponents)
+    {
+        if ($this->dataName() === 'devBranch') {
+            $this->markTestIncomplete('Not implemented');
```
Comment (martinjunger) [inline tests/Common/ComponentsTest.php:67]: Tady je to zajimavy, devBranch pise, ze to neni implementovany 🤔

**Comment by martinjunger** [comment]
Takhle vypada vysledek (zatim):

Screenshot at Sep 23 17-24-47

Screenshot at Sep 23 17-25-52
)

File: tests/Common/ComponentsTest.php [hunk 5]
```diff
@@ -38,10 +38,39 @@ public function setUp()
         }
     }
 
-    public function testGetComponentDetail()
+    public function componentsProvider()
     {
+        yield 'defaultBranch' => [
+            function (\Keboola\StorageApi\Client $client) {
```
Comment (martinjunger) [inline tests/Common/ComponentsTest.php:44]: Aby to nemusela byt anonymni funkce, ktere se predava argument `$this->_client`, tak by opet bylo potreba zmenit `$this->_client` na **public**. Je to OK?

File: src/Keboola/StorageApi/Components.php (hunk 1 above)
Comment (tomasfejfar) [reply src/Keboola/StorageApi/Components.php:31]: IMHO je to špatný směr. Z components by sis klienta tahat neměl.

File: tests/Common/ComponentsTest.php [hunk 6]
```diff
@@ -38,10 +38,39 @@ public function setUp()
...
+                $branch = $devBranch->createBranch($branchName);
+                return new \Keboola\StorageApi\Components($this->getBranchAwareDefaultClient($branch['id']));
+            }
+        ];
+    }
+
+    /**
+     * @dataProvider componentsProvider
+     */
+    public function testGetComponentDetail(callable $getComponents)
```
Comment (tomasfejfar) [inline tests/Common/ComponentsTest.php:64]:
```suggestion
    /**
     * @dataProvider componentsProvider
     */
    public function testGetComponentDetail(Client $client)
```

A pak v kódu, když potřebuješ výhybku, tak `$client instanceof BranchAwareClient`

Comment (tomasfejfar) [reply tests/Common/ComponentsTest.php:64]: To z nějakého důvodu nejde?

File: tests/Common/ComponentsTest.php [hunk 7]
```diff
@@ -464,31 +517,47 @@ public function testComponentConfigRestrictionsForReadOnlyUser()
...
+    /**
+     * @dataProvider componentsProvider
+     */
+    public function testConfigurationNameShouldBeRequired(callable $getComponents)
     {
+        /** @var \Keboola\StorageApi\Components $components */
+        $components = $getComponents($this->_client);
         try {
-            $this->_client->apiPost('components/wr-db/configs', []);
+            $components->getClient()->apiPost('components/wr-db/configs', []);
```
Comment (tomasfejfar) [inline tests/Common/ComponentsTest.php:528]: tohle bys nemusel řešit kdybys z dataprovideru dostal $client

Comment (tomasfejfar) [reply tests/Common/ComponentsTest.php:528]: plus volání klienta přímo je edge case, který by klidně dataprovider nemusel podporovat IMHO

File: src/Keboola/StorageApi/Components.php (hunk 1 above)
Comment (martinjunger) [reply src/Keboola/StorageApi/Components.php]: Jo, zruseno.

File: tests/Common/ComponentsTest.php (hunk 5 above)
Comment (martinjunger) [reply tests/Common/ComponentsTest.php]: Tak ne, `$client` se tam musi dostat az z testu, protoze `dataProvider` se spousti z jine classy a jeste pred `setUp` metodou, takze `$this->_client` je `null`.

File: tests/Common/ComponentsTest.php (hunk 7 above)
Comment (martinjunger) [reply tests/Common/ComponentsTest.php]: Jo, toho jsem si nevsim, upraveno, ze vraci Client.

File: tests/Common/ComponentsTest.php (hunk 6 above)
Comment (martinjunger) [reply tests/Common/ComponentsTest.php]: Instance Clienta neni v DataProvideru dostupna - viz nize. Proto je potreba predat `$this` a nastavit `devBranch` clienta az za behu.
> All data providers are executed before both the call to the setUpBeforeClass static method and the first call to the setUp method. Because of that you can't access any variables you create there from within a data provider. This is required in order for PHPUnit to be able to compute the total number of tests.
> https://phpunit.de/manual/6.5/en/writing-tests-for-phpunit.html#:~:text=a%20depending%20test.-,Note,-All%20data%20providers

File: tests/Common/ComponentsTest.php [hunk 8]
```diff
@@ -38,10 +40,39 @@ public function setUp()
...
+        yield 'defaultBranch' => [
+            function (ComponentsTest $that) {
+                return $that->_client;
+            }
+        ];
+
+        yield 'devBranch' => [
+            function (ComponentsTest $that) {
+                $providedToken = $that->_client->verifyToken();
+                $branchName = __CLASS__ . '\\' . $that->getName(false) . '\\' . $that->dataName() . '\\' . $providedToken['id'];
```
Comment (jirkasemmler) [inline tests/Common/ComponentsTest.php:54]: byl bych pro `implode`

File: tests/Common/ComponentsTest.php [hunk 9]
```diff
@@ -38,10 +40,39 @@ public function setUp()
         }
     }
 
-    public function testGetComponentDetail()
+    public function componentsProvider()
```
Comment (jirkasemmler) [inline tests/Common/ComponentsTest.php:43]: prosim nejaky obsahly komentar vysvetlujici co a proc tady provadis

File: tests/Common/ComponentsTest.php [hunk 10]
```diff
@@ -464,31 +515,50 @@ public function testComponentConfigRestrictionsForReadOnlyUser()
...
 
-    public function testConfigurationNameShouldBeRequired()
+    /**
+     * @dataProvider componentsProvider
+     */
+    public function testConfigurationNameShouldBeRequired(callable $getClient)
     {
+        /** @var Client $client */
+        $client = $getClient($this);
+        $components = new \Keboola\StorageApi\Components($client);
```
Comment (jirkasemmler) [inline tests/Common/ComponentsTest.php:525]: na co jsou tady ty $components? ani v constructoru to nedela zadnou logiku (tak nejak vsude)

File: tests/Common/ComponentsTest.php (hunk 2 above)
Comment (jirkasemmler) [reply tests/Common/ComponentsTest.php:1318]: chapu, ale v tom pripade bych je dal jako skipped nebo neco. Potrebujeme mergnout neco, co nam by default nebude padat

**Review by jirkasemmler — COMMENTED** [review]
par malickosti, prosim o doplneni  . padaji mi celkem 4 testy, shodne s tvymi, ale ty mas jeden navic

image

File: tests/Common/ComponentsTest.php (hunk 2 above)
Comment (martinjunger) [reply tests/Common/ComponentsTest.php:1318]: jj, tyhle jeste prozkoumavam. Potom jim dam bud skipped nebo incomplete.

File: tests/Common/ComponentsTest.php (hunk 8 above)
Comment (martinjunger) [reply tests/Common/ComponentsTest.php:54]: jj, muze byt.

**Comment by martinjunger** [comment]
> par malickosti, prosim o doplneni . padaji mi celkem 4 testy, shodne s tvymi, ale ty mas jeden navic

to je tim, ze nemam nastavenyho `read only` token, takze mi pada navic `testComponentConfigRestrictionsForReadOnlyUser`.

File: tests/Common/ComponentsTest.php (hunk 10 above)
Comment (martinjunger) [reply tests/Common/ComponentsTest.php]: ha, to je pozustatek z predchozi iterace. Dik

File: tests/Common/ComponentsTest.php (hunk 9 above)
Comment (martinjunger) [reply tests/Common/ComponentsTest.php:53]: done

File: tests/Common/ComponentsTest.php (hunk 2 above)
Comment (martinjunger) [reply tests/Common/ComponentsTest.php]: zkoumam dal, ale dal jsem je jako `incomplete`

**Comment by martinjunger** [comment]
Pridal jsem provider do dalsich 2 test suit:
- ConfigurationRowsSortOrderTest - pokryto vsechno - vsechny OK
- ConfigurationRowTest - pokryto 3/6
  - 2 - pouziva specificky vytvorene klienty - viz TODO
  - 1 - pouziva jiz jiny `dataProvider` - jak tam pridat jednoduse druhy?
 
Trochu jsem jeste poladil:
- ComponentsTest - pokryto skoro vsechno - vetsina OK, ale
  - 6 - pouziva specificky vytvorene klienty - zatim jsem preskocil viz `TODO`
  - 4 - vraci jine nez ocekavana vysledky 🔴 
  - 1 - vraci Not implemented 🤔 
  - 6 - skipped, nepodporovano pro dev branche

**Comment by martinjunger** [comment]
Aktualne vsechny testy projdou jako ✅  success nebo 🚫  skipped/ignored, takze na testingu to nezpusobi zadne problemy.

Screenshot 2021-10-01 at 12 27 20

Screenshot 2021-10-01 at 12 37 12

Screenshot 2021-10-01 at 12 34 25

File: tests/Common/ComponentsTest.php [hunk 11]
```diff
@@ -412,6 +444,7 @@ public function testComponentConfigCreate()
         $this->assertEquals('some desc', $configuration['description']);
     }
 
+    // TODO vice jinych klientu
```
[the remark this answers is not in the document]
Comment (martinjunger) [reply tests/Common/ComponentsTest.php:447]: jj, nejspis to nejak takhle pujde, budu resit v dalsim tiketu.
